src/java.base/share/classes/jdk/internal/classfile/ClassSignature.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.classfile;
+
+import java.util.List;
+import jdk.internal.classfile.impl.SignaturesImpl;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Models the generic signature of a class file, as defined by JVMS 4.7.9.
+ */
+public sealed interface ClassSignature
```

Comment:
It is a bit odd to have Signature and XYZSignature in the API with the latter not extending the former. I think I get what the API is aiming for though - e.g. Signature is for modelling low-level "portions" of the signature attributes, whereas ClassSignature, MethodSignature and friends are there to model the signature attribute attached to a class, method, etc.

Reply:
The confusion come from simplified name. Signature probably should be called JavaTypeSignature according to the spec. ClassSignature and MethodSignature could not extend it would not respect the reality, each of them are completely different species. Signature (or JavaTypeSignature) on the other side has many sub-types.

Reply:
I think you mean `TypeSignature` from JLS 4.3.4 ? If so, I get it. To me it looks as if "Signature" really means "TypeSignature" - then it would make sense as to why `MethodSignature::arguments` returns a `List<Signature>` (e.g. because it should be `List<TypeSignature>`, as in the spec).

Also, from a modelling perspective, I see other problems too - in the JVMS, type signatures are defined as follows:

```
TypeSignature:
    FieldTypeSignature
    BaseType
```    
And:

```
FieldTypeSignature:
    ClassTypeSignature
    ArrayTypeSignature
    TypeVariableSignature 
```

So I'd roughly expect a type with 4 subclasses (type, then class <: type, array <: type, tvar <: type).

But the Signature class currently has other subclasses too - such as `ThrowableSig` - which is really only an (optional) part of method signatures. And, similarly, it also has `TypeParam`, which is a part of method/class signature - but is _not_ a type signature per se.

Summing up, it seems to me that the naming issue (which is not that important) hides a bigger modelling issue.

Reply:
Class `Signature` (aka `JavaTypeSignature`), all subclasses, `MethodSignature` and `ClassSignature` are designed according to [JVMS 4.7.9.1 Signatures](https://docs.oracle.com/javase/specs/jvms/se19/html/jvms-4.html#jvms-4.7.9.1)

Reply:
**Specification:**
```
MethodSignature:
[TypeParameters] ( {JavaTypeSignature} ) Result {ThrowsSignature}

Result:
JavaTypeSignature 
VoidDescriptor

ThrowsSignature:
^ ClassTypeSignature 
^ TypeVariableSignature
```


**Reflect in API mapping:**
```
    public sealed interface ClassTypeSig
            extends RefTypeSig, ThrowableSig
```
and
```
    public sealed interface TypeVarSig
            extends RefTypeSig, ThrowableSig
```
and
```
    /**
     * @return method signature
     * @param typeParameters signatures for the type parameters
     * @param exceptions sigantures for the exceptions
     * @param result signature for the return type
     * @param arguments signatures for the method arguments
     */
    public static MethodSignature of(List<Signature.TypeParam> typeParameters,
                                     List<Signature.ThrowableSig> exceptions,
                                     Signature result,
                                     Signature... arguments) {
```

`Signature.ThrowableSig` is a `Signature` and it is a common super of `ClassTypeSig` and `ClassTypeSig`.

Reply:
`TypeParam` is not a signature, because it simply is not a signature.
Per spec:
```
TypeParameter:
Identifier ClassBound {InterfaceBound}
```

Reply:
> `Signature.ThrowableSig` is a `Signature` and it is a common super of `ClassTypeSig` and `TypeVarSig`.

I really don't follow here. ThrowableSig is a piece of a method signature, which starts with "^" and is followed by either a class or a typevar signature. You can never encounter it from the toplevel JavaTypeSignature production. It's ok (as I have said) to have a ThrowableSig element in the API to model the production, but that element should not be a subtype of Signature (at least not if Signature, as you claimed is meant to model the JavaTypeSignature production). That is, there's no "is a" relationship between JavaTypeSignature and ThrowableSig (at least not that I can see when looking at the productions).

Reply:
OK, I see your point now, I'll fix it.
Thanks

Reply:
During the fix I found the definition that `ThrowableSig` (as a super-set of `ClassTypeSig` and `TypeVarSig`) IS `Signature` - that fact is critical for processing.
For example when `MethodSignature` is serialized ThrowableSig::signatureString` is critical.
Or `ClassRemapper` depends on the inheritence:
```
signature.throwableSignatures().stream().map(this::mapSignature).toList()
```
however `mapSignature` could not be applied on `ThrowableSig` when it does not inherit from `Signature`.

Reply:
I think I also see where the current hierarchy comes from - e.g. while `ThrowableSig` is a part of a method signature, it is indeed used by the production to specify a set of signatures that belong to that set. Thanks for the patience.

Reply:
No problem, I had to remind myself all the reasons why does it look like Rubik's cube :)